WIP: LX-116 Converts Video XModule to an XBlock

pomegranited opened the pull request:
WIP: Initial move of video_module code into xblock_builtin/xblock_video

* Video XBlock loads in LMS with original Video XModule OLX and separate video definition files.
* Videos are no longer exported to a separate `videos` definition files; they're inlined in the vertical.
* Both `video` and legacy `videoalpha` types supported (see Demo Course)

**JIRA tickets**: [OSPR-2830](https://openedx.atlassian.net/browse/OSPR-2830) LX-116

**Dependencies**: 

* https://github.com/edx/XBlock/pull/419

**Screenshots**: N/A -- no UI changes expected.

**Sandbox URL**
* LMS: https://pr19303.sandbox.opencraft.hosting/
* Studio: https://studio-pr19303.sandbox.opencraft.hosting/

(f22baec -- redeploying)

**Note** We're experiencing a bug with authentication for our sandboxes based on the latest master, and so to use the sandbox, you may need to work around this issue by logging into the Django Admin instead:
* LMS: https://pr19303.sandbox.opencraft.hosting/admin
* Studio: https://studio-pr19303.sandbox.opencraft.hosting/admin

**Testing instructions**

Setup notes:
* You'll need to run the following to use this branch on your devstack:
   ```
   make lms-static
   make studio-static 
   ```

Testing:
1. Import the [edx Demo course](https://github.com/edx/edx-demo-course/releases/tag/open-release%2Fhawthorn.2).
1.  Ensure that the videos in the Demo course are visible, specifically:
    * [Introduction: Video and Sequences](https://preview-pr19303.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+DemoX+2018_1/courseware/d8a6192ade314473a78242dfeedfbf5b/edx_introduction/1?activate_block_id=block-v1%3AOpenCraft%2BDemoX%2B2018_1%2Btype%40vertical%2Bblock%40vertical_0270f6de40fc) -- `videoalpha` block type
    * [ Videos on edX](https://pr19303.sandbox.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/courseware/interactive_demonstrations/19a30717eff543078a5d94ae9d6c18a5/1?activate_block_id=block-v1%3AedX%2BDemoX%2BDemo_Course%2Btype%40vertical%2Bblock%40867dddb6f55d410caaa9c1eb9c6743ec) -- `video` block type, shows transcript.
1. Enable completion tracking by activating the `completion.enable_completion_tracking` waffle flag.
    Ensure that watching videos to the end triggers a `publish_completion` event (404s, see notes)
1. Export the Demo course and compare the export with the original [edx Demo course](https://github.com/edx/edx-demo-course/releases/tag/open-release%2Fhawthorn.2).
   
**Author Notes & Concerns**
1. Viewing, adding, and editing video blocks in Studio is not working yet.
1. Subtitles and transcripts have not been tested extensively, so some issues may remain with imported/exported courses.
1. This initial PR is a prototype, and so most code has been moved with minimal modifications.
  Code refactoring and cleanup and test fixes are expected before this PR will be ready for approval.
1. Completion event handling is not working yet.

**Reviewers**
- [ ] @bradenmacdonald 
- [ ] edX reviewer[s] TBD -- FYI @ormsbee

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-2830](https://openedx.atlassian.net/browse/OSPR-2830) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

pomegranited commented on common/lib/xmodule/xmodule/fields.py:
```diff
@@ -299,3 +302,24 @@ def from_json(self, value):
+    def from_json(self, value):
+        # Empty strings sometimes come in from OLX
+        if isinstance(value, basestring) and not value:
+            value = self._default
```
This is a temporary hack to deal with some OLX with `html5_sources=""`.  I can fix this on `XBlock.fields.List`, unless there's another way to sanitize this value before its stored on the XBlock?

pomegranited replied:
@bradenmacdonald  Yep, cool, I've submitted https://github.com/edx/XBlock/pull/419 and dropped this change in favour of using that branch in b55a82ee6e9047445ef8b19f9cab93d47210d478.

pomegranited commented on openedx/core/lib/xblock_builtin/xblock_video/xblock_video/video_xfields.py:
```diff
@@ -4,6 +4,8 @@
 from xblock.fields import Scope, String, Float, Boolean, List, Dict, DateTime
+
+# TODO JV: Move to XBlock if only used by Video?
 from xmodule.fields import RelativeTime
```
Note to self: Can move this field into `XBlock.fields` as part of https://github.com/edx/XBlock/pull/419 too, since there doesn't appear to be anything else in edx-platform using that class, I can also move that field type into `XBlock.fields`. 

Will need to migrate the [RelativeTime JS editor](https://github.com/edx/edx-platform/blob/a439d5164c07e4695181b15244e8e5c7681421c4/cms/static/js/views/metadata.js#L364) and [tests](https://github.com/edx/edx-platform/blob/a439d5164c07e4695181b15244e8e5c7681421c4/cms/static/js/spec/views/metadata_edit_spec.js#L518) to [use the editor logic in xblock-utils](https://github.com/edx/xblock-utils/blob/18fcd70ff8eed6af17d41b16cf66a8161c9fb51b/xblockutils/studio_editable.py#L96-L108).

pomegranited commented on .gitignore:
```diff
@@ -100,7 +100,7 @@ lms/static/css/
 cms/static/css/
 common/static/common/js/vendor/
-common/static/common/css/vendor/
+common/static/common/css
```
@bradenmacdonald Those files are in [`common/static/css`](https://github.com/edx/edx-platform/tree/master/common/static/css), not [`common/static/common/css`](https://github.com/edx/edx-platform/tree/master/common/static/common) -- this ignored dir is not checked into the repo.

pomegranited commented on lms/envs/common.py:
```diff
@@ -1444,6 +1444,44 @@ def _make_locale_paths(settings):
 ]
 
+video_js = (
+    rooted_glob(COMMON_ROOT / 'static', 'js/video/time.js') +
```
Good point!  Fixed with e752e02

pomegranited replied:
Update: had to revert this change, see https://github.com/edx/edx-platform/pull/19303#issuecomment-448428450

pomegranited commented on openedx/core/lib/xblock_builtin/xblock_video/xblock_video/__init__.py:
```diff
@@ -0,0 +1,147 @@
+class VideoXBlock(
+        VideoFields,
+        VideoMixin,
+        VideoDescriptor,
```
Yep, I was aiming to reduce the diff, so 3ef5059 renames `VideoDescriptor` to `VideoEditingMixins`.

pomegranited commented on openedx/core/lib/xblock_builtin/xblock_video/xblock_video/__init__.py:
```diff
@@ -0,0 +1,147 @@
+        If no external data sources are found (file in "video" folder), it is exactly equivalent to base method
+        XBlock.parse_xml. Otherwise this method parses file in "video" folder (known as definition_xml) and updates fields accordingly.
+        """
+        block = super(VideoDescriptor, cls).parse_xml(node, runtime, keys, id_generator)
```
That was unintentional, fixed with https://github.com/edx/edx-platform/commit/3ef5059b1e7310e03d2016936f3d56cbb84de4c2#diff-7433df0404d1462c997ed373e01284f4R101

pomegranited commented on openedx/core/lib/xblock_builtin/xblock_video/xblock_video/__init__.py:
```diff
@@ -0,0 +1,147 @@
+        VideoStudentViewHandlers,
+        VideoStudioViewHandlers,
+        XBlock,
+        XmlParserMixin):
```
Fixed with https://github.com/edx/edx-platform/pull/19303/commits/0f6a46b1dc96fec24e394eb4072d99825b1814f9#diff-7433df0404d1462c997ed373e01284f4R32

pomegranited commented on openedx/core/lib/xblock_builtin/xblock_video/xblock_video/__init__.py:
```diff
@@ -0,0 +1,147 @@
+        folder in OLX to store fields and children.
+
+        If no external data sources are found (file in "video" folder), it is exactly equivalent to base method
+        XBlock.parse_xml. Otherwise this method parses file in "video" folder (known as definition_xml) and updates fields accordingly.
```
Fixed with https://github.com/edx/edx-platform/pull/19303/commits/0f6a46b1dc96fec24e394eb4072d99825b1814f9#diff-7433df0404d1462c997ed373e01284f4R59

pomegranited replied:
I [copied this logic from `xblock_discussion`](https://github.com/edx/edx-platform/blob/master/openedx/core/lib/xblock_builtin/xblock_discussion/xblock_discussion/__init__.py#L260), but now that I look at this closer, it is suprising that it's not handled already by the [XmlParserMixin.parse_xml](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/xml_module.py#L325) method.

Perhaps it`s failing the [`is_pointer`](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/xml_module.py#L351) check, and so isn't loading the `video` content automatically?  Will need to look into this further, so added a [FIXME note](https://github.com/edx/edx-platform/pull/19303/files#diff-7433df0404d1462c997ed373e01284f4R80) to the code.

pomegranited commented on openedx/core/lib/xblock_builtin/xblock_video/xblock_video/video_handlers.py:
```diff
@@ -277,99 +275,127 @@ def make_transcript_http_response(content, filename, language, content_type, add
+        Dispatches, (HTTP GET):
+            /translation/[language_id]
+            /download
+            /available_translations/
```
Fixed with https://github.com/edx/edx-platform/pull/19303/commits/0f6a46b1dc96fec24e394eb4072d99825b1814f9#diff-82b3724f87854c364a5bfc2f5c02141cL376

pomegranited commented on openedx/core/lib/xblock_builtin/xblock_video/xblock_video/video_module.py:
```diff
@@ -402,25 +361,24 @@ def get_html(self):
-class VideoDescriptor(VideoFields, VideoTranscriptsMixin, VideoStudioViewHandlers,
-                      TabsEditingDescriptor, EmptyDataRawDescriptor, LicenseMixin):
+class VideoDescriptor(LicenseMixin):
+    # FIXME re-add base class functionality: TabsEditingDescriptor, EmptyDataRawDescriptor
```
Noted with `FIXME` at https://github.com/edx/edx-platform/pull/19303/commits/0f6a46b1dc96fec24e394eb4072d99825b1814f9#diff-82b3724f87854c364a5bfc2f5c02141cR203

pomegranited commented on openedx/core/lib/xblock_builtin/xblock_video/xblock_video/__init__.py:
```diff
@@ -0,0 +1,147 @@
+        return fragment
+
+    @staticmethod
+    def css_dependencies():
```
Fixed with https://github.com/edx/edx-platform/pull/19303/commits/0f6a46b1dc96fec24e394eb4072d99825b1814f9#diff-7433df0404d1462c997ed373e01284f4R59

pomegranited commented on common/static/js/video/08_video_speed_control.js:
```diff
@@ -1,8 +1,8 @@
     define(
-'video/08_video_speed_control.js', [
-    'video/00_iterator.js',
+'js/video/08_video_speed_control.js', [
```
Ah yep, that was a SASS issue that I didn't detect, since it didn't throw a compile error.  Fixed with 20160ac .

pomegranited commented:
Thanks for your review @bradenmacdonald !  I've addressed your comments (or at least noted that they need addressing in a subsequent ticket), and am building a new sandbox now to verify the asset and XML parser changes.

Once the sandbox is up and verified, is it ok for me to rebase on the latest master to bring in your merged https://github.com/edx/edx-platform/pull/19260 ?

bradenmacdonald commented:
@pomegranited Changes look good! Yes, please rebase :) Thanks!

Since this PR isn't merge ready I won't be giving a formal approval, but we can consider this done for the scope of the prototype ticket.

bradenmacdonald commented:
When trying out your latest rebased version via the LX frontend, I'm seeing a requireJS error in the JS console:

```
Script error for "js/video/time.js", needed by: js/video/03_video_player.js, js/video/09_video_caption.js
```

It might be an issue on my side, but I wasn't seeing it before rebasing.

pomegranited commented:
@bradenmacdonald  That script error sounds like something that could have been introduced by d236669036ef71d582e0f2939d7902193af6689e, when we changed the JS file globbing.  Let me know if you see it again?

pomegranited commented:
Hi @ormsbee !  @symbolist or I will start updating this PR in a couple of weeks to provide Studio editing support, and to fix the remaining issues with this conversion.

Is there anything in particular that you want this PR to address, or changes to our approach that you'd recommend?  And is there any major work planned on the Video XModule that we need to plan for? CC @bradenmacdonald

ormsbee commented:
@pomegranited: Approach looks good. No other major work happening on this module that I'm aware of.

@pdpinch: I seem to recall there were transcript related compatibility issues in the past that bit us. Do you remember the details of that? I don't remember what the resolution to that was or if there are adequate test cases to cover that now.

pdpinch commented:
There definitely have been issues, but I've had trouble tracking down the particulars. Based on recollection, I think there a coupe of issues -- but I don't know if they intersect with this PR. 

1. There was a change in mid 2018 to the transcript representation in OLX. There's a page in confluence about the change, but it's hard to find.

2. Separately, we've discovered that the code for handling export for transcripts is overly sensitive to errors. If a file cannot be found, the export fails altogether. I don't know how much of an edge case this is, but we wasted a lot of time chasing it down. 

> I seem to recall there were transcript related compatibility issues in the past that bit us. Do you remember the details of that? I don't remember what the resolution to that was or if there are adequate test cases to cover that now.

Colin-Fredericks commented:
Two questions so far from the HarvardX side on this one:

1. Is there still going to be a separate video/ folder, or is that going away? There's a comment from @bradenmacdonald about that back in December, but I can't tell which way it was resolved. Inlining would break our archiving scripts, as per the discussion currently happening on the [Problem upgrade PR](https://github.com/edx/edx-platform/pull/19071/files/39e28b1be9df3e31d4708b0177dc90657cd72be0..8630ebe2e1041f93a517e0f482a8c62e57fcdfbe). If you've got a quick solution here that keeps the folder, perhaps that will be useful in the problem PR?
2. The video DOM element (by which I mean the div with the .video class) currently has a number of useful things in it, including data attributes with javascript functions that can be used to control the video. The code that I use for pop-up questions and links depends on that - this is the code I [gave a talk about](https://docs.google.com/presentation/d/1k-i6vthLpxLbaS_vlTeINTh3YB6cWnIh7k6LqArmNVM/edit?usp=sharing) at the last OpenEdX in Montreal. Is this going to disappear?

Thanks!

pomegranited commented:
Hi @Colin-Fredericks !

> 1. Is there still going to be a separate video/ folder, or is that going away?

Maybe.  I'm looking into this as part of https://github.com/edx/edx-platform/pull/19071, and whatever we find, we'll use similar logic to do the same thing here.  I don't think it's going to be a quick solution though!

> 2. The video DOM element (by which I mean the div with the .video class) currently has a number of useful things in it, including data attributes with javascript functions that can be used to control the video. The code that I use for pop-up questions and links depends on that - this is the code I [gave a talk about](https://docs.google.com/presentation/d/1k-i6vthLpxLbaS_vlTeINTh3YB6cWnIh7k6LqArmNVM/edit?usp=sharing) at the last OpenEdX in Montreal. Is this going to disappear?

We have no intention of changing how the Video XBlock is displayed or used, so if you notice anything critical that's missing, please let us know.  You are welcome to use the sandboxes linked in this PR description for testing this change.

From the current prototype [Video XBlock deployed on the sandbox](https://pr19303.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+DemoX+2018_1/courseware/d8a6192ade314473a78242dfeedfbf5b/edx_introduction/1?activate_block_id=block-v1%3AOpenCraft%2BDemoX%2B2018_1%2Btype%40html%2Bblock%40030e35c4756a4ddc8d40b95fbbfff4d4), I can see that the `div.video` still contains these data attributes:
* `data-metadata`, e.g. 
   ```javascript
   {"savedVideoPosition": 0.0, "ytMetadataUrl": "https://www.googleapis.com/youtube/v3/videos/", "sources": ["https://s3.amazonaws.com/edx-course-videos/edx-edx101/EDXSPCPJSP13-H010000_100.mp4"], "duration": 0.0, "saveStateEnabled": true, "speed": null, "generalSpeed": 1.0, "end": 0.0, "ytTestTimeout": 1500, "completionPercentage": 0.95, "transcriptLanguages": {"en": "English"}, "start": 0.0, "ytKey": null, "poster": null, "recordedYoutubeIsAvailable": true, "publishCompletionUrl": "/courses/course-v1:OpenCraft+DemoX+2018_1/xblock/block-v1:OpenCraft+DemoX+2018_1+type@videoalpha+block@0b9e39477cf34507a7a48f74be381fdd/handler/publish_completion", "ytApiUrl": "https://www.youtube.com/iframe_api", "prioritizeHls": false, "autoAdvance": false, "saveStateUrl": "/courses/course-v1:OpenCraft+DemoX+2018_1/xblock/block-v1:OpenCraft+DemoX+2018_1+type@videoalpha+block@0b9e39477cf34507a7a48f74be381fdd/handler/save_user_state", "transcriptAvailableTranslationsUrl": "/courses/course-v1:OpenCraft+DemoX+2018_1/xblock/block-v1:OpenCraft+DemoX+2018_1+type@videoalpha+block@0b9e39477cf34507a7a48f74be381fdd/handler/available_translations", "transcriptLanguage": "en", "showCaptions": "true", "autohideHtml5": false, "streams": "1.00:b7xgknqkQk8", "completionEnabled": true, "transcriptTranslationUrl": "/courses/course-v1:OpenCraft+DemoX+2018_1/xblock/block-v1:OpenCraft+DemoX+2018_1+type@videoalpha+block@0b9e39477cf34507a7a48f74be381fdd/handler/translation/__lang__", "captionDataDir": null, "autoplay": false}
   ```
* `data-bumper-metadata`
* `data-autoadvance-enabled`
* `data-poster`

Colin-Fredericks commented:
@pomegranited 
> Maybe. I'm looking into this as part of #19071, and whatever we find, we'll use similar logic to do the same thing here. I don't think it's going to be a quick solution though!
> 

Fair'nuff.

> We have no intention of changing how the Video XBlock is displayed or used, so if you notice anything critical that's missing, please let us know. You are welcome to use the sandboxes linked in this PR description for testing this change.
> 

Great, thanks! seems to still be there and working well.

pomegranited commented on openedx/core/lib/xblock_builtin/xblock_video/xblock_video/video_xfields.py:
```diff
@@ -12,7 +14,7 @@
 
 
 class VideoFields(object):
```
For the CapaXBlock PR, we determined that XBlock fields should live in the base class, else they are treated differently than expected.  cf https://github.com/open-craft/edx-platform/pull/145/commits/0c120c58631fd18283761d437a30b51bbc13edc7

pomegranited replied:
Also, if any fields need to be exported as separate XML nodes (e.g. `transcripts`), then use [`xml_node=True`](https://github.com/edx/XBlock/blob/7cb80f5dec610cb577c49ca7ee6c6c844fd33c3d/xblock/fields.py#L307-L308).

pomegranited commented:
Superceded by https://github.com/edx/edx-platform/pull/20384

edx-webhook commented:
@pomegranited Even though your pull request wasn’t merged, please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/19303&entry.1484655318=https://github.com/pomegranited&entry.752974735=2018-11-22+08:09&entry.1917517419=2019-05-06+17:41&entry.2133058324=No) so we can improve your experience in the future.
